src/main/kotlin/io.javalin/Request.kt

```diff
@@ -0,0 +1,203 @@
+/*
+ * Javalin - https://javalin.io
+ * Copyright 2017 David Åse
+ * Licensed under Apache 2.0: https://github.com/tipsy/javalin/blob/master/LICENSE
+ */
+
+package io.javalin
+
+import io.javalin.core.util.RequestUtil
+import io.javalin.core.util.Util
+import io.javalin.translator.json.Jackson
+import java.io.IOException
+import java.util.*
+import java.util.concurrent.CompletableFuture
+import javax.servlet.http.Cookie
+import javax.servlet.http.HttpServletRequest
+
+class Request(private val servletRequest: HttpServletRequest,
+              private val paramMap: Map<String, String>, // cache (different for each handler)
+              private val splatList: List<String>) { // cache (different for each handler)
+
+    private var passedToNextHandler: Boolean = false
+
+    fun unwrap(): HttpServletRequest {
+        return servletRequest
+    }
+
+    @FunctionalInterface
+    interface AsyncHandler {
+        fun handle(): CompletableFuture<Void>
+    }
```

Comment:
Functional interfaces could be replaced by typealiases 

```kotlin
typealias AsyncHandler = () -> CompletableFuture<Void>
```

The only caveats are that it can't be nested in a class and also sometimes they had problems with return types involving Kotlin `Unit` when are used on Java